### PR TITLE
Improved type definitions for the observable query

### PR DESCRIPTION
**Pull request by excitement-engineer**
The [Query](https://github.com/apollographql/react-apollo/blob/master/src/Query.tsx) component in react-apollo improved upon the typescript definitions provided by apollo-client. 

I am now porting these improvements back to apollo-client so that all users can benefit from these improvements and so that react-apollo can simply import the type definitions directly from apollo-client.

This PR makes introduce generic types for the data `TData` and variables `TVariables` in the `ObservableQuery` class.

### CODE REVIEW

File: packages/apollo-client/src/core/ObservableQuery.ts [hunk 1]
```diff
@@ -421,17 +430,14 @@ export class ObservableQuery<T> extends Observable<ApolloQueryResult<T>> {
    *
    */
   public setVariables(
-    variables: any,
+    variables?: TVariables,
```
Comment (excitement-engineer) [inline packages/apollo-client/src/core/ObservableQuery.ts]: Question: Is this supposed to be nullable. The type was `any` before so anything was allowed. If I make the variables nonnull here then I get typescript errors.

File: packages/apollo-client/src/core/ObservableQuery.ts [hunk 2]
```diff
@@ -33,18 +33,21 @@ export type ApolloCurrentResult<T> = {
...
+  TVariables = OperationVariables
+> {
   updateQuery: (
-    previousQueryResult: { [key: string]: any },
+    previousQueryResult: TData,
     options: {
-      fetchMoreResult?: { [key: string]: any };
-      variables: { [key: string]: any };
+      fetchMoreResult?: TData;
+      variables?: TVariables;
```
Comment (excitement-engineer) [inline packages/apollo-client/src/core/ObservableQuery.ts]: Question: Is this supposed to be nullable? It wasn't before but typescript is giving me errors if I make this nonnull.

**Comment by excitement-engineer** [comment]
@jbaxleyiii @leoasis

File: packages/apollo-client/src/core/ObservableQuery.ts (hunk 2 above)
Comment (jbaxleyiii) [reply packages/apollo-client/src/core/ObservableQuery.ts]: @excitement-engineer what errors did you get?

Comment (jbaxleyiii) [reply packages/apollo-client/src/core/ObservableQuery.ts]: but, yes variables can be null here

File: packages/apollo-client/src/core/ObservableQuery.ts (hunk 1 above)
Comment (jbaxleyiii) [reply packages/apollo-client/src/core/ObservableQuery.ts]: hmm, I don't the the intention is for it to be nullable here. What issues did you run into?

**Comment by jbaxleyiii** [comment]
oh @excitement-engineer could you add a changelog here as well? Thanks!

File: packages/apollo-utilities/src/__tests__/transform.ts [hunk 3]
```diff
@@ -982,4 +982,5 @@ describe('getDirectivesFromDocument', () => {
       const doc = getDirectivesFromDocument([{ name: 'client' }], query, true);
       expect(print(doc)).toBe(print(expected));
     });
+  });
```
Comment (excitement-engineer) [inline packages/apollo-utilities/src/__tests__/transform.ts]: This was missing in apollo-client. This was causing the prettier pre-commit hook to fail...

File: packages/apollo-client/src/core/ObservableQuery.ts [hunk 4]
```diff
@@ -426,10 +435,10 @@ export class ObservableQuery<T> extends Observable<ApolloQueryResult<T>> {
...
    */
   public setVariables(
-    variables: any,
+    variables: TVariables,
     tryFetch: boolean = false,
     fetchResults = true,
-  ): Promise<ApolloQueryResult<T>> {
+  ): Promise<ApolloQueryResult<TData>> {
     // since setVariables restarts the subscription, we reset the tornDown status
     this.isTornDown = false;
```
Comment (excitement-engineer) [inline packages/apollo-client/src/core/ObservableQuery.ts]: In the line below here we are doing: 

```js
const newVariables = variables ? variables : this.variables;
```

So I am wondering if it isn't possible to make the type `variables?: TVariables`?

File: packages/apollo-client/src/core/ObservableQuery.ts [hunk 5]
```diff
@@ -399,7 +408,7 @@ export class ObservableQuery<T> extends Observable<ApolloQueryResult<T>> {
       false;
 
     return this.setVariables(
-      this.options.variables,
+      this.options.variables as TVariables,
```
Comment (excitement-engineer) [inline packages/apollo-client/src/core/ObservableQuery.ts]: This is the issue I ran into with `setVariables` having a non-null TVariables type. `this.options.variables` is nullable here so I had to make an explicit cast to get typescript to work

**Pull request merged**